[WIP] - tfe.keras - logistic regression

yanndupis opened the pull request:
Hey,

#540 
This is a WIP PR to support simple models (e.g. logistic regression) with tfe.keras interface. 

There are several remaining items:
- [ ] `.fit()` expects private input right now but it should work with numpy array as well. In a separate PR we could support `tf.data.Dataset` by creating a TFE version of it.
- [ ] support accuracy
- [ ] potentially use callbacks to modularize `fit` step and track loss, accuracy etc. https://keras.io/callbacks/
- [ ] add doctrings to optimizers, loss etc. 
- [ ] in binary cross entropy support `from_logit` parameter
- [ ] refactor gradient tests for Dense and sigmoid layer to make it cleaner.

mortendahl commented on tf_encrypted/keras/losses.py:
```diff
@@ -0,0 +1,53 @@
+
+  batch_size = y_true.shape.as_list()[0]
+  batch_size_inv = 1 / batch_size
+  out = y_true * get_protocol().log(y_pred)
```
maybe we can't do it now but eventually (ie after refactoring) we should get rid of the `get_protocol` call and simply use `tfe.log(..)` instead (which will find the right protocol)

yanndupis replied:
Agreed

mortendahl commented on tf_encrypted/keras/losses.py:
```diff
@@ -0,0 +1,53 @@
+  batch_size = y_true.shape.as_list()[0]
+  batch_size_inv = 1 / batch_size
+  out = y_true * get_protocol().log(y_pred)
+  out += (y_true.negative() + 1) * get_protocol().log(y_pred.negative() + 1)
```
does `1 - y_true` work?

yanndupis replied:
It does, nice! I didn't know 1 will be lifted this way as well when the other tensor is a private tensor.

mortendahl commented on examples/notebooks/keras-training/common.py:
```diff
@@ -0,0 +1,179 @@
+import tf_encrypted as tfe
+
+
+class LogisticRegression:
```
maybe it's time we move this guy into its own directory, say `tfe.keras.applications` ([to match](https://www.tensorflow.org/api_docs/python/tf/keras/applications))

yanndupis replied:
Not sure. I think this repo is focusing on pre-trained models: https://keras.io/applications/

mortendahl replied:
makes sense 👍

mortendahl commented on examples/notebooks/keras-training/data.py:
```diff
@@ -0,0 +1,44 @@
+"""Dummy data preparation for logistic regression training."""
```
is this file used anywhere?

yanndupis replied:
Fixed - deleted, thanks

mortendahl commented on tf_encrypted/keras/models/sequential.py:
```diff
@@ -106,6 +110,33 @@ def layers(self):
+
+  def fit_batch(self, x, y):
+    y_pred = self.call(x)
+    dy = self._loss.grad(y, y_pred)
```
what happens if you compiled with `loss=None`?

yanndupis replied:
I think it will crash :)

mortendahl commented on tf_encrypted/keras/models/sequential.py:
```diff
@@ -106,6 +110,33 @@ def layers(self):
+
+  def fit(self, x, y, epochs=1, steps_per_epoch=1):
+    for e in range(epochs):
+      print('Epoch {}/{}'.format(e + 1, epochs))
```
use logger instead (if anything)

yanndupis replied:
Fixed -  I used logger instead. But it looks quite ugly lol. Maybe there is the option to remove the pink color at least

mortendahl commented on tf_encrypted/keras/optimizers.py:
```diff
@@ -0,0 +1,24 @@
+    self.lr = lr
+
+  def apply_gradients(self, var, grad):
+    sess = K.get_session()
```
wouldn't this give you a `tf.Session` instead of `tfe.Session`?

yanndupis replied:
Fixed - good catch

yanndupis commented on tf_encrypted/keras/backend.py:
```diff
@@ -20,9 +20,9 @@ def get_session(op_input_list=()):
+    if isinstance(default_session, tfe.Session):
       return default_session
-    if not isinstance(default_session, tfe.Session()):
+    if not isinstance(default_session, tfe.Session):
```
Fixed

yanndupis commented on tf_encrypted/keras/__init__.py:
```diff
@@ -14,4 +15,5 @@
     'layers',
     'models',
     'Sequential',
+    'losses'
```
Fixed

yanndupis commented on tf_encrypted/keras/layers/activation.py:
```diff
@@ -18,12 +18,20 @@ class Activation(Layer):
+
+  def backward(self, d_y):
+    grad_weights = []
+    d_x = d_y * self.activation_deriv(self._layer_output)
```
Fixed

yanndupis commented on tf_encrypted/keras/losses.py:
```diff
@@ -0,0 +1,53 @@
+               fn,
+               **kwargs):
+
+    self.fn = fn
```
Fixed

yanndupis commented:
Thank you @mortendahl  for your review! I have addressed most of the comments (noted Fixed). Not sure I will have much time to work on this next week. So feel free to push some code if needed to get it merge. Thanks again.

mortendahl commented:
Alright, I can take over for now, thanks @yanndupis!

mortendahl commented:
A larger issue is how to handle gradients. The current code allows eg activation functions to be given, but doesn't handle derivatives correctly. Not sure what the best approach is here, maybe follow the pattern of `tf.custom_gradient` where two functions are returned?

yanndupis commented:
Hey @mortendahl , what do you mean by "eg activation functions to be given, but doesn't handle derivatives correctly.". 

one thing that the code is not handling is if the activation is specified in the Dense layer (e.g. tfe.keras.layers.Dense(1, activation='sigmoid'), it will work only if you do `tfe.keras.layers.Dense(1) then `tfe.keras.layers.Activation('sigmoid')

Are you referring to something else?

yanndupis commented:
@mortendahl, we can use this issue: https://github.com/tf-encrypted/tf-encrypted/issues/540
